nmigen/vendor/xilinx_s7.py

```diff
@@ -0,0 +1,254 @@
+from abc import abstractproperty
+
+from ..hdl.ast import *
+from ..hdl.dsl import *
+from ..hdl.ir import *
+from ..build import *
+
+
+__all__ = ["XilinxS7Platform"]
+
+
+class XilinxS7Platform(TemplatedPlatform):
+    """
+    Required tools:
+        * ``vivado``
+
+    Available overrides:
+        * ``script_after_read``: inserts commands after ``read_xdc`` in Tcl script.
+        * ``script_after_synth``: inserts commands after ``synth_design`` in Tcl script.
+        * ``script_after_place``: inserts commands after ``place_design`` in Tcl script.
+        * ``script_after_route``: inserts commands after ``route_design`` in Tcl script.
+        * ``script_before_bitstream``: inserts commands before ``write_bitstream`` in Tcl script.
+        * ``script_after_bitstream``: inserts commands after ``write_bitstream`` in Tcl script.
+        * ``additional_constraints``: inserts commands in XDC file.
+        * ``vivado_opts``: adds extra options for Vivado.
+
+    Build products:
+        * ``{{name}}.log``: Vivado log.
+        * ``{{name}}_timing_synth.rpt``: Vivado report.
+        * ``{{name}}_utilization_hierarchical_synth.rpt``: Vivado report.
+        * ``{{name}}_utilization_synth.rpt``: Vivado report.
+        * ``{{name}}_utilization_hierarchical_place.rpt``: Vivado report.
+        * ``{{name}}_utilization_place.rpt``: Vivado report.
+        * ``{{name}}_io.rpt``: Vivado report.
+        * ``{{name}}_control_sets.rpt``: Vivado report.
+        * ``{{name}}_clock_utilization.rpt``:  Vivado report.
+        * ``{{name}}_route_status.rpt``: Vivado report.
+        * ``{{name}}_drc.rpt``: Vivado report.
+        * ``{{name}}_timing.rpt``: Vivado report.
+        * ``{{name}}_power.rpt``: Vivado report.
+        * ``{{name}}_route.dcp``: Vivado design checkpoint.
+        * ``{{name}}.bit``: binary bitstream.
+    """
+
+    device = abstractproperty()
```

Comment:
Is this property going to contain combined device, package and speed grade? It would be more future-proof (and to a degree more convenient) if those were split, for several reasons. First, that makes it easier to branch on device, package, etc in nMigen code portable across a family that wants to change its behavior accordingly. Second, other tools, like nextpnr, generally use separate parameters for these.